### PR TITLE
enhancement #861: created the endpoint for receiving all students

### LINKED ISSUE #861: Improve the endpoint for receiving students
We have an endpoint to receive students by **group**. First, abstract this endpoint by adding the group id to the query (so that you can get all the students). However, we need to expand its functionality by adding **filtering by groups and group roles** _(multichoice)_. Also you need to add a **search**.

image

### PR DESCRIPTION
**Pull request by f-ogoog**
- created the endpoint for receiving all students
- added AllStudentsPipe.ts, SimpleStudentResponse, QueryAllStudentDTO.ts

closes #861

### CODE REVIEW

File: src/v2/api/dtos/QueryAllStudentDTO.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  @IsOptional()
+    search?: string;
+
+  @ApiPropertyOptional({
+    description: 'Sorting parameters',
+    enum: SortQGSParam,
+    default: SortQGSParam.LAST_NAME,
+  })
+  @IsOptional()
+  @IsIn([SortQGSParam.FIRST_NAME, SortQGSParam.LAST_NAME, SortQGSParam.MIDDLE_NAME], validationOptionsMsg('Sort should be an enum'))
```
Comment (ArtemiiLiakh) [inline src/v2/api/dtos/QueryAllStudentDTO.ts]: it's better to use IsEnum

File: src/v2/api/dtos/QueryAllStudentDTO.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    enum: ['asc', 'desc'],
+    default: 'asc',
+  })
+    order?: 'asc' | 'desc';
+
+  @ApiPropertyOptional({
+    description: 'Array of groups ids',
+  })
+  @IsArray()
+  @IsOptional(validationOptionsMsg('Groups should be an array'))
```
Comment (ArtemiiLiakh) [inline src/v2/api/dtos/QueryAllStudentDTO.ts:41]: it's better to use the validationMessage in IsArray

File: src/v2/api/dtos/QueryAllStudentDTO.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    groups?: string[];
+
+  @ApiPropertyOptional({
+    description: 'Array of students roles',
+    enum: [RoleName.STUDENT, RoleName.MODERATOR, RoleName.CAPTAIN],
+    type: [RoleName.STUDENT, RoleName.MODERATOR, RoleName.CAPTAIN],
+  })
+  @IsArray(validationOptionsMsg('Roles should be an array'))
+  @IsEnum([RoleName.STUDENT, RoleName.MODERATOR, RoleName.CAPTAIN], { each: true, message: 'Each element of roles should be an enum' })
```
Comment (ArtemiiLiakh) [inline src/v2/api/dtos/QueryAllStudentDTO.ts]: It's better to create a new enum and use [NewEnum] or something like this (don't forget to change other rows with this array)
also make the code more readable

File: src/v2/api/responses/StudentResponse.ts [hunk 4]
```diff
@@ -21,6 +21,39 @@ export class BaseStudentResponse {
...
+
+  @ApiProperty({
+    description: 'Student\'s state',
+    enum: State,
+  })
+    state: State;
+
+  @ApiProperty({
+    description: 'Student\'s roles',
+    enum: [RoleName.STUDENT, RoleName.MODERATOR, RoleName.CAPTAIN],
```
Comment (ArtemiiLiakh) [inline src/v2/api/responses/StudentResponse.ts]: also use a new enum there

File: src/v2/database/DatabaseUtils.ts [hunk 5]
```diff
@@ -23,6 +23,28 @@ export class DatabaseUtils {
     };
   }
 
+  static getStrictSearch (search: string | number, field: string) {
```
Comment (ArtemLyah) [inline src/v2/database/DatabaseUtils.ts:26]: Where is this used?

**Review by ArtemiiLiakh — APPROVED** [review]

File: src/v2/api/responses/StudentResponse.ts [hunk 6]
```diff
@@ -21,6 +22,39 @@ export class BaseStudentResponse {
...
+    description: 'Student\'s state',
+    enum: State,
+  })
+    state: State;
+
+  @ApiProperty({
+    description: 'Student\'s group role',
+    enum: GroupRoles,
+  })
+    role: string;
```
Comment (illiakroshka) [inline src/v2/api/responses/StudentResponse.ts:41]: In swagger it shows like number 
image

**Review by illiakroshka — CHANGES_REQUESTED** [review]

**Comment by illiakroshka** [comment]
It looks like roles does't work. Sends back all students istead of students with moderator or admin role

image

**Review by illiakroshka — APPROVED** [review]

File: src/v2/api/pipes/AllStudentsPipe.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+@Injectable()
+export class AllStudentsPipe implements PipeTransform<QueryAllStudentDTO, Promise<QueryAllStudentDTO>> {
+  constructor (
+    private readonly groupByIdPipe: GroupByIdPipe,
+  ) {}
+
+  async transform (query: QueryAllStudentDTO): Promise<QueryAllStudentDTO> {
+    if (query.groups) {
+      await Promise.all(query.groups?.map((groupId) => this.groupByIdPipe.transform(groupId)));
```
Comment (fokaaas) [inline src/v2/api/pipes/AllStudentsPipe.ts]: use mapAsync in utils

File: src/v2/security/PERMISSION.ts [hunk 8]
```diff
@@ -92,4 +92,6 @@ export enum PERMISSION {
   TEACHERS_$TEACHERID_CONTACTS_CREATE = 'teachers.$teacherId.contacts.create',
   TEACHERS_$TEACHERID_CONTACTS_UPDATE = 'teachers.$teacherId.contacts.update',
   TEACHERS_$TEACHERID_CONTACTS_DELETE = 'teachers.$teacherId.contacts.delete',
+
+  ADMIN_STUDENTS_GET = 'admin.students.get'
```
Comment (fokaaas) [inline src/v2/security/PERMISSION.ts]: just students.get

**Review by fokaaas — CHANGES_REQUESTED** [review]

**Review by fokaaas — APPROVED** [review]

**Pull request merged**